src/frontend-samples/query-volume-app/VolumeQueryApp.tsx

```diff
@@ -0,0 +1,179 @@
+    }
+    // Create a box for the ClipVector.
+    const block: ClipShape = ClipShape.createBlock(range, range.isAlmostZeroZ ? ClipMaskXYZRangePlanes.XAndY : ClipMaskXYZRangePlanes.All, false, false);
+    // Create the ClipVector
+    const clip: ClipVector = ClipVector.createEmpty();
+    // Add the box to the Clipvector and set it in the ScreenViewport.
+    clip.appendReference(block);
+    // Call enableClipVolume to ensure all clip flags are properly set
+    ViewClipTool.enableClipVolume(vp);
+    // Turning off the clipping feature.
+    vp.view.viewFlags.clipVolume = isClippingOn === undefined ? false : isClippingOn;
+    vp.view.setViewClip(clip);
+    VolumeQueryApp.addDecorators(vp);
+  }
+
+  /* Method for adding decorators to the viewport */
+  public static addDecorators(vp: ScreenViewport) {
+    // Create a clip decorator. Selecting the clip decoration to immediately show the handles is the default.
+    const vcdp: ViewClipDecorationProvider = ViewClipDecorationProvider.create();
+    // Default behavior is to hide the decorators on deselect. We want to keep the decorators showing in this example.
+    vcdp.clearDecorationOnDeselect = false;
+    vcdp.showDecoration(vp);
+    // The decorators require the SelectTool being active.
+    IModelApp.toolAdmin.startDefaultTool();
+  }
+
+  /* Getting the range of currect box */
+  public static getRangeOfBox(vp: ScreenViewport): Range3d | undefined {
+    const viewClip = vp.view.getViewClip();
+    if (viewClip !== undefined) {
+      return ClipUtilities.rangeOfClipperIntersectionWithRange(
+        viewClip,
+        vp.view.computeFitRange(),
+      );
+    }
+
+    return undefined;
+  }
+
+  /* Getting all elements that are spatially represented in iModel*/
+  public static async getAllSpacialElements(vp: ScreenViewport): Promise<PhysicalElement[]> {
+    const esqlQuery = "SELECT ECInstanceId, ECClassId FROM BisCore.PhysicalElement";
+    const elementsAsync = vp.iModel.query(esqlQuery);
+    const elements: PhysicalElement[] = [];
+    for await (const element of elementsAsync) {
+      elements.push({ id: element.id, className: element.className, name: "" });
+    }
+
+    const presentationProvider = new PresentationLabelsProvider({ imodel: vp.iModel });
+    const packedInstancesKeys: InstanceKey[][] = [];
+    const instanceKeys: InstanceKey[] = elements.map((element) => {
+      return { className: element.className.replace(".", ":"), id: element.id };
+    });
+
+    /* Sorting elements into packages. Instead of one 3000 length array I will have 3 x 1000 lenght arrays */
+    const packsOfInstanceKeys = Math.floor(instanceKeys.length / 1000);
+    for (let i = 0; i <= packsOfInstanceKeys; i++) {
+      if (i !== packsOfInstanceKeys) {
+        packedInstancesKeys.push(instanceKeys.slice(i * 1000, (i + 1) * 1000));
+      } else {
+        packedInstancesKeys.push(instanceKeys.slice(i * 1000, instanceKeys.length + 1));
+      }
+    }
+
+    /* Getting elements names */
```

Comment:
Maybe worth a comment here to say, the reason why you created the packages was to avoid making a single query for a very large number of labels.

Also... why are you getting the labels for all the elements if you are only going to show the first 100?

Reply:
Well I am getting all the labels because I have interface PhysicalElement, that has id, name, className as its properties. The program works like this(just changed it for performance) - when you select iModel, all physicalElements are loaded and put into VolumeQueryUI class property with ids, names and classNames. Then when you change the volume of the box and click apply, you don't load all the elements again. You take them from VolumeQueryUI property and pass them to API so it would show which ones are inside/outside/overlap. In other words, firstly you load all the PhysicalElements, and then you work with them(list them or do other operations). So getting only 100 elements names would not work in this scenario. Also the main reason why the slice is being done is not because we load too much elements and it takes time, but because representing 1000 elements in list takes ~1 sec(you can feel it when changing what elements you want to see - inside/outside/overlap).